\documentclass[11pt]{article}

\usepackage{times}

\usepackage{amsmath}
\usepackage{amsthm}
\usepackage{amssymb}
\usepackage{mathabx} 
\usepackage{graphicx}
\usepackage{color} 
\usepackage{setspace} 
\usepackage{rotating}
\usepackage{natbib}

\usepackage{multirow}
\usepackage{xspace}
\usepackage{lscape}
%\usepackage{cite}
\usepackage{xr}
%\externaldocument{poly-div-supp}

% for continued lists
\usepackage{enumitem}


\usepackage{hyperref}
\urlstyle{rm}
\hypersetup{
  colorlinks,
  urlcolor=blue,
  linkcolor=black,
  citecolor=black
}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in

\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

\definecolor{gray}{rgb}{0.3, 0.3, 0.4}
% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\newcommand{\smallCom}[1]{\marginpar{\tiny{#1}}}

\newtheorem{claim}{Claim}
\newtheorem{lemma}{Lemma}
\newtheorem{corollary}{Corollary}
\newtheorem{definition}{Definition}


\newcommand{\eqdef}{\stackrel{\Delta}{=}}		
\DeclareMathOperator*{\argmin}{\arg\!\min}

\newcommand{\vect}[1]{\boldsymbol{\mathbf{#1}}}
\newcommand{\ld}{\mathcal{L}}
\newcommand{\ignore}[1]{}

\newcommand{\mcref}{\textsc{McRef}\xspace}

\newcommand{\E}{\mathbb{E}}
\newcommand{\X}{\vect{X}}
\newcommand{\M}{\mathcal{M}}
\newcommand{\Tr}{\mathcal{T}}
\newcommand{\B}{\vect{B}}
\newcommand{\Y}{\vect{Y}}
\newcommand{\G}{\vect{G}}
\newcommand{\T}{\vect{\Theta}}
\newcommand{\GT}{\G\T}
\newcommand{\Mref}{\M_{ref}}
\newcommand{\Pref}{\widetilde{P}}
\newcommand{\rbf}{\text{BF}}
%\newcommand{\hbf}{\text{BF}}
\newcommand{\Om}{\Omega}

\newcommand{\taus}{\vect\tau}
\newcommand{\thetas}{\vect\theta}
\newcommand{\migs}{\vect{m}}


\newcommand{\GTref}{\widetilde{\GT}}
\newcommand{\Gref}{\widetilde{\G}}
\newcommand{\Tref}{\widetilde{\T}}
\newcommand{\thref}{\widetilde{\thetas}}
\newcommand{\taref}{\widetilde{\taus}}
\newcommand{\migref}{\widetilde{\migs}}
\newcommand{\Z}{\vect{Z}}
\newcommand{\Zref}{\widetilde{\Z}}
\newcommand{\Omref}{\widetilde{\Om}}

\newcommand{\Fext}{F_{Z}}

\newcommand{\troot}{\theta_{root}}
\newcommand{\thcomb}{\theta_{comb}}
\newcommand{\tacomb}{\tau_{comb}}

\newcommand{\Gc}{\G_c}
\newcommand{\Gm}{\G_m}

\def\comb{\rotatebox[origin=c]{90}{$\exists$}}
\newcommand{\Mcomb}{\M_{\comb}}
\newcommand{\Gcomb}{\G_{\comb}}
\newcommand{\Tcomb}{\T_{\comb}}
\newcommand{\tmin}{\tau_{\text{min}}}

% Two lines from genres
\def\@cite#1#2{(#1\if@tempswa , #2\fi)}
\def\@biblabel#1{}
\makeatother

\pagestyle{myheadings}
%% ** EDIT HERE **
\markright{A New Bayesian Method for Comparing Demographic Models}

\begin{document}

\begin{titlepage}

%\title{Generative Probabilistic Model for Detecting Selection on
%Dispersed Genomic Elements from Polymorphism and Divergence }
\title{A New Bayesian Method for Comparing Demographic Models}


\author{Ron Visbord$^1$, Ilan Gronau$^{1}$}

%\date{ }
\maketitle

\begin{footnotesize}
\begin{center}
$^1$Efi Arazi School of Computer Science, Herzliya Interdisciplinary Center (IDC), Herzliya 46150, Israel
\end{center}
\end{footnotesize}

\vspace{1in}

\begin{tabular}{lp{4.5in}}
{\bf Submission type:}& Research Article
\vspace{1ex}\\
{\bf Keywords:}& 
Molecular evolution, population phylogenomics, demography inference, model comparison
\vspace{1ex}\\
{\bf Running Head:}&Bayesian Method for Comparing Demographic Models
\vspace{1ex}\\ 
{\bf Corresponding Author:}&
\begin{minipage}[t]{4in}
 Ilan Gronau\\
 Efi Arazi School of Computer Science\\
 The Herzliya Interdisciplinary Center\\
 P.O.B. 167 Herzliya, 46150 Israel\\
 Phone: +972-9-952-7907\\
 Fax: +972-9-956-8604\\
 Email: ilan.gronau@idc.ac.il
\end{minipage}
\end{tabular}

\thispagestyle{empty}
\end{titlepage}

\doublespacing

\section*{Abstract}

% basic abstract for method. Need to amp up.

Demography inference has emerged as a fundamental task when analyzing population genomic data.
%
Bayesian methods for demography inference provide a powerful and flexible framework for estimating
parameter values in complex demographic models, but they do not provide means for assessing the fit
of the assumed model to the data.
%
This is because Bayes factors, which are used to measure relative model fit, are notoriously difficult to estimate
robustly for complex probabilistic models.
%
We present here a new approach to compare the fit of different demographic models to genomic data
by directly estimating Bayes factors with respect to a given reference model.
%
Similar to standard estimators of the Bayes factor, our estimate is based on importance sampling.
%
However, unlike standard estimators, careful selection of the reference model allows us to produce
robust and accurate comparisons between models.
%
We introduce the theory behind estimation of relative Bayes factors, and demonstrate its usefulness in a variety of settings.
%
Importantly, this new approach is easily applicable to many existing Bayesian demography inference methods.
%
We provide an implementation of relative Bayes factors for the Generalized Phylogenetic Coalescent Sampler (G-PhoCS),
and apply it to simulated and real genomic data.


%\thispagestyle{empty}
%\clearpage
\setcounter{page}{1}

\section*{Introduction}

% review previous work and demo inf methods
-- TBA --



\section*{Preliminaries: demographic models and Bayesian inference}

A probabilistic demographic model $\M$ uses a parameterized demographic history to define a probability distribution over observed genomic data $\X$.
%
The structural components of $\M$, which we assume are fixed, consist of a population phylogeny $\Tr$ and a collection
of migration bands $B$ that indicate ordered pairs of populations between which gene flow is allowed.
%
The free parameters of $\M$, denoted here by $\T$, consist of divergence times, $\taus=\{\tau_p:p \text{ is an ancestral population of } \Tr\}$,
effective population sizes, $\thetas=\{\theta_p: p \text{ is a population in } \Tr\}$, and migration rates, $\migs=\{m_b:b \in B\}$.
%
All model parameters are scaled by mutation rate.
%
The model $\M$ is thus defined by specifying the structural components $(\Tr,B)$ and a prior distribution over the free parameters of the model $P(\T|\M)$.
%
The conditional probability distribution for the observed genomic data, $P(\X|\M,\T)$, is defined by standard models for
molecular evolution and population genetics (e.g., \cite{JUKECANT69,KING82A}).
%
The objective of demography inference methods is to infer values for $\T$ that have high joint probability with the data:
$P(\X,\T|\M)=P(\T|\M)P(\X|\M,\T)$.

%
Because the conditional probability $P(\X|\M,\T)$ does not typically have a closed-form expression, an increasingly popular approach for
inference is to introduce additional hidden variables $\G$, which represent genealogical relationships
between the sampled individuals.
%
The benefit of this is that given the genealogical information, the data $\X$ becomes independent of the model $\M$ and parameters $\T$,
and the likelihood can be expressed as a product of three efficiently computable terms:
%
%
\begin{equation}\label{eq:likelihood}
 P(\X,\G,\T|\M) ~=~ P(\T|\M) P(\G|\M,\T) P(\X|\G)~.
\end{equation}
%
%

This joint probability function may be used by a Markov chain Monte Carlo (MCMC) algorithm to generate a sample of the model parameters
together with the genealogies according to a probability distribution approximating the posterior, $P(\G,\T|\M,\X)$.
%
Consequently, the sampled parameter values have high joint probability with the data.
%
A major advantage of this approach to inference is that it is extremely flexible and can be applied to a wide range of demographic models and
different types of genomic data.
%
For simplicity, we will consider here a model for sequence data at  short unlinked loci, in which case 
$\G$ contains the information on the local tree in each locus, and distinct loci are assumed to be independent (e.g., \cite{NIELWAKE01,RANNYANG03,GRONETAL11}).
%
However, the framework we describe here is quite general and can be extended to other types of data and more complex demographic models.

%The data, $\X$, may consist of long contigs, in which case the genealogical information in $\G$ explicitly represent genetic recombination 
%along the analyzed sequence (see, e.g., \cite{RASMETAL14}).
%

\section*{Methods}


\subsection*{Estimating data likelihood via importance sampling}

A fundamental limitation of Bayesian demography inference methods is that they do not directly produce reliable measures of model fit.
%
% In particular, the likelihood values recorded by the MCMC algorithm, $P(\X,\G,\T|\M)$, depend on the representation of hidden
% genealogies in the model and are thus not directly comparable across models.
%
Model fit is best captured by the marginal data likelihood, $P(\X|\M)$, whose computation involves integration over the space of unknown parameter values and genealogical relationships,
denoted jointly by $\GT$.
%
This high-dimensional integral may be approximated via importance sampling using a collection of instances $\{\GT^{(i)}\}$ sampled via MCMC conditioned on $\X$ and $\M$.
%
The approximation is established by expressing the inverse of the likelihood as an expected value under the posterior distribution of $\GT$ given $\M$ and $\X$:
%
%
\begin{small}
\begin{align}
%\frac{1}{\hbf(\M|\X)} ~~~ \triangleq~~
\frac{1}{P(\X|\M)} ~~~
&=~~ \frac{\int P(\GT|\M)d\GT}{P(\X|\M)} \notag \\ %
&=~~ \int \frac{P(\GT|\M)}{P(\X|\M)} \frac{P(\X,\GT|\M)}{P(\X,\GT|\M)}  d\GT \notag \\ %
&=~~ \int \frac{P(\GT,\X |\M)}{P(\X|\M)} ~\bigg/ \frac{P(\X,\GT|\M)}{P(\GT|\M)}  d\GT \notag \\ %
&=~~ \int \frac{P(\GT|\M,\X)}{P(\X|\M,\GT)} d\GT \notag \\ %
&=~~ \int \frac{1}{P(\X|\G)}P(\GT|\M,\X) d\GT  \notag \\ %
&=~~ \E_{\GT|\M,\X } \left[\frac{1}{P(\X|\G)}\right] \notag\\  %~.\label{eq:is_harmonic}\\
%\notag \\ %
%\frac{1}{P(\X|\M)}
&\approx~~ \frac{1}{N} \sum_{i=1}^{N}\frac{1}{P(\X|\G^{(i)})} ~. \label{eq:harmonic}
\end{align}
\end{small}

This {\em harmonic mean estimator} is straightforward and can be applied in a very general setting, but its practical use is often limited due to very high variance
of the inverse likelihood, $1/P(\X|\G)$.
%
This high variance means that only models with very different levels of fit may be compared reliably via harmonic mean estimators of $P(\X|\M)$.
%
The main objective of the approach we propose next is to correlate the sensitivity of model comparison with the level of similarity between the models being compared.
%

\subsection*{Relative Bayes factors}

We propose here an alternative way to evaluate the fit of model $\M$ by estimating its likelihood relative to some
reference model $\Mref$. 
%
As before, assume a collection $\{\GT^{(i)}\}$ sampled via MCMC according to an approximate posterior probability distribution $P(\GT|\M,\X)$.
%
We wish to use these MCMC samples to estimate the {\em Bayes factor of $\M$ relative to $\Mref$}, defined as the ratio $P(\X|\M) / P(\X|\Mref)$.
%
The Bayes factor can be estimated by running an additional MCMC for $\Mref$ and taking the ratio of the two harmonc-mean estimates for $P(\X|\M)$ and $P(\X|\Mref)$.
%
However, in some cases the relative Bayes factor may be estimated directly from $\{\GT^{(i)}\}$ without the need for an additional MCMC for $\Mref$.
%
This is done by connecting the models $\M$ and $\Mref$ via a conditional distribution over the the hidden variables of $\M$, $\Pref(\GT|\Mref)$,
which satisfies the following two requirements:
%
%
\begin{small}
\begin{align}
&P(\X|\Mref) ~~=~~ \int  \Pref(\GT|\Mref)\ P(\X|\G)\ d\GT \label{eq:pref_integral}\\
&P(\GT|\M,\X)=0 ~~\Rightarrow~~ \Pref(\GT|\Mref)=0 \label{eq:pref_support}
\end{align}
\end{small}
%
%


The \emph{model pairing conditional distribution}, $\Pref(\GT|\Mref)$, plays a key role in our estimator for the relative Bayes factor.
%
The special notation $\Pref$ indicates that this probability function is not naturally defined by either
$\M$ or $\Mref$, and there will typically be some degree of freedom associated with its specification.
%
Given a model-pairing conditional distribution, the relative Bayes factor  may be expressed as an expected value under the posterior distribution of $\GT$ given $\M$ and $\X$,
implying the following approximation:
%
%
\begin{small}
\begin{align}
\frac{1}{\rbf(\M:\Mref|\X)} ~~~ \triangleq ~~ \frac{P(\X|\Mref)}{P(\X|\M)}
&=~~ \frac{\int  \Pref(\GT|\Mref)\ P(\X|\G)\ d\GT}{P(\X|\M)} \label{eq:pref1} \\ %
&=~~ \int \frac{\Pref(\GT|\Mref)\ P(\X|\G) }{P(\X|\M)} \ \frac{P(\GT|\M, \X)}{P(\GT|\M, \X)}  d\GT \label{eq:pref2} \\ %
&=~~ \int \frac{\Pref(\GT|\Mref)\ P(\X|\G)\ }{P(\X,\GT|\M)} P(\GT|\M, \X)  d\GT \notag \\ %
&=~~ \int \frac{\Pref(\GT|\Mref) }{P(\GT|\M)} P(\GT|\M, \X)  d\GT  \label{eq:data_cancel}\\ %
&=~~ \E_{\GT|\M,X } \left[\frac{\Pref(\GT|\Mref) }{P(\GT|\M)}\right]~.\notag \\
&\approx~~ \frac{1}{N} \sum_{i=1}^{N}\frac{\Pref(\GT^{(i)}|\Mref) }{P(\GT^{(i)}|\M)} ~.\label{eq:rbf}
\end{align}
\end{small}
%
%

Note that the condition of Equation \ref{eq:pref_integral} implies the equality in Equation \ref{eq:pref1},
and the condition of Equation \ref{eq:pref_support} guarantees no division-by-zero in Equation \ref{eq:pref2}.
%
Interestingly, the contribution of the data to the likelihood cancels out in Equation \ref{eq:data_cancel} (because it is equal in both models).
%
Thus the ratio used for estimation, ${\Pref(\GT|\Mref) }/{P(\GT|\M)}$, is not a direct function of the data ($\X$),
and the data affects the estimate only through its influence the sampled instances $\{\GT^{(i)}\}$.
%
Importantly, the variance of this ratio, which we refer to as the {\em relative Bayes factor (RBF) ratio},
%is potentially smaller than the variance of the inverse likelihoods used in Equation \ref{eq:harmonic}.
%
%In particular, this variance
depends on the definition of the model-pairing conditional, $\Pref$, and it will
typically decrease as $\M$ and $\Mref$ become more similar.
For instance, in the trivial case where $\Mref=M$, we can define $\Pref(\GT|\Mref)=P(\GT|\M)$ and the RBF ratio becomes
1 for all instances $\{\GT^{(i)}\}$.
%
This is the key advantage of direct estimation of the Bayes factor, when compared to estimation via harmonic mean, and
realizing this advantage requires construction of an effective model-pairing conditional distribution for $\M$ and $\Mref$.
%, and this is not possible for all pairs of models. 
%
The following sections present specific constructions for $\Pref$ in a series of cases.


\subsection*{The null reference model $\M_0$}

We start by considering a simple case where $\M$ is a demographic model with no migration bands and $\Mref$
is the simplest possible model with a single population of constant size, $\theta_0$.
%
We refer to this simple one-parameter model as the {\em null reference model}, $\M_0$.
%
% Construction of the model-pairing conditional is done in {three steps}.
%
%\textbf
{The first step} of constructing a model-pairing conditional for the two models is to identify a mapping $F$ from the space
of hidden variables in $\M$ to the space of hidden variables in $\M_0$.
%, allowing us to express the likelihood under $\M_0$ as an integral over hidden variables in $\M$.
%
In our case, denote by $\Gref$ and $\Tref$ the hidden variables of $\M_0$.
%
Because both $\M$ and $\M_0$ have no migration bands, then we may assume that the genealogical information used
by both models is the same, implying a natural one-to-one mapping between $\G$ and $\Gref$ (the implications of migration are discussed in the next section).
%
%However, while $\Tref=\{\theta_0\}$, $\T$ will typically consist of more parameters: $\T=\{\tau_p\}\cup\{\theta_p\}$.
%
A mapping between $\T=(\taus,\thetas)$ and $\Tref=(\theta_0)$ can be defined by selecting one of the population size parameters in $\T$
to be associated with $\theta_0$. This can be the size of the root population, $\troot$, or any other population
that we expect to best represent the single population in $\M_0$.
%
The model pairing conditional is obtained by applying this mapping and extending it to the unmapped hidden variables, $~ \Z=(\taus,\thetas\setminus \{\troot\})$,
%= \{\tau_p\}\cup\{\theta_p\}_{p\neq root}$,
with the use of a conditional distribution, $\Pref(\Z|\GT\setminus\Z)$:
%
%
\begin{small}
\begin{equation}
 \Pref(\GT|\M_0)  ~~=~~
 P(\theta_0=\troot|\M_0)\ P(\Gref=\G|\M_0,\theta_0=\troot)\ \Pref(\Z|\G,\troot)   ~ .\label{eq:pref_null}
\end{equation}
\end{small}
%
%
The model-pairing condition of Equation \ref{eq:pref_integral} is thus established, regardless of how $\Pref(\Z|\G,\troot)$ is defined:
%
%
\begin{small}
\begin{align}
P(\X|\M_0)
&=~~ \int P(\Tref|\M_0)\ P(\Gref|\M_0,\Tref)\ P(\X|\Gref)\   d\Gref d\Tref  \notag \\ %
&=~~ \int P(\theta_0=\troot|\M_0)\ P(\Gref=\G|\M_0,\theta_0=\troot)\ P(\X|\G)\  d\G d\troot \notag \\ 
&=~~ \int P(\theta_0=\troot|\M_0)\ P(\Gref=\G|\M_0,\theta_0=\troot)\ P(\X|\G)\
\left( \int \Pref(\Z|\G,\troot)\ d\Z \right) d\G d\troot \notag \\ 
%
&=~~ \int P(\theta_0=\troot|\M_0)\ P`(\Gref=\G|\M_0,\theta_0=\troot)\ \Pref(\Z|\G,\troot)\ P(\X|\G)\ d\GT \notag \\ 
&=~~ \int \Pref(\GT|\M_0)\ P(\X|\G)\ d\GT ~. \label{eq:likelihood_null} %
\end{align}
\end{small}

% Alternative definitions of the model-pairing function can be achieved by selecting a
% different set of hidden variables from $\GT$ to map to $\Gref\Tref$, and/or altering the conditional
% distribution over the remaining hidden variables, $\Pref(\Z|\GT\setminus\Z)$. We designate this conditional probability
% function with $\Pref$ to indicate that it is not strictly implied by any of the models of interest.

%
%The model pairing conditional is thus shaped based on how we .

We are left to construct $\Pref(\Z|\G,\troot)$ so that it ensures the model-pairing condition of Equation \ref{eq:pref_support},
and we wish to use the remaining degree of freedom to minimize the variance of the RBF ratio.
%
Equation \ref{eq:pref_support} is guaranteed by constricting $\Pref(\Z|\G,\troot)$ to have zero values whenever $P(\G,\troot,\Z|\M,\X)=0$.
%
Among the unmapped variables $\Z=(\taus,\thetas\setminus \{\troot\})$, the population size parameters $\thetas\setminus \{\troot\}$ do not 
pose any restrictions on the mapped variables $\G,\troot$. This means that Equations \ref{eq:pref_support} is guaranteed regardless of how their marginal distribution is defined.
%
We thus define their conditional probability distribution according to their prior probability in $\M$, to cancel out terms in the RBF ratio and potentially reduce its variance.
%
%
%
%\begin{equation}
% \Pref(\Z|\G,\troot) ~~=~~ \Pref(\{\theta_p\}_{p\neq root},\{\tau_p\}|\G) ~~=~~ 
% \Pref(\{\tau_p\}|\G) \ \prod_{p\neq root} P(\theta_p|\M)\  ~.\label{eq:cond_tau}
%\end{equation}
%
%
\begin{small}
\begin{align}
\frac{\Pref(\GT|\M_0) }{P(\GT|\M)}
&=~~ \frac{ P(\theta_0=\troot|\M_0) ~ P(\G|\M_0,\theta_0=\troot) ~ \Pref(\Z|\G,\troot)} {P(\GT|\M)} \notag \\
&=~~ \frac{ P(\G|\M_0,\theta_0=\troot) }{ P(\G|\M,\T)}~ 
     \frac{ P(\theta_0=\troot|\M_0) \prod_{p\neq \troot}\Pref(\theta_p|\G,\troot) }{P(\troot|\M)\prod_{p\neq \troot}P(\theta_p|\M)}~
     \frac{ \Pref(\taus|\G,\thetas)}{P(\taus|\M)} \notag \\
&=~~ \frac{ P(\G|\M_0,\theta_0=\troot) }{ P(\G|\M,\T)}~ 
     \frac{ P(\theta_0=\troot|\M_0)}{P(\troot|\M)}\
     \frac{ \Pref(\taus|\G,\thetas)}{P(\taus|\M)} ~. \label{eq:rbf_null}
\end{align}
\end{small}

Note that if we assume that $\M$ and $\M_0$ use the same prior distribution over $\theta_{root}$ and $\theta_0$ (resp.),
then the middle term in Equation \ref{eq:rbf_null} also cancels out.
%
We cannot similarly define $\Pref(\taus|\G,\thetas)=P(\taus|\M)$, because this may lead to conflicts between divergence times and coalescence times in $\G$, which result in violation of
the model-pairing condition of Equation \ref{eq:pref_support}.
%
Such conflicts occur when a divergence time $\tau_p$ is deeper than the most recent common ancestor
in $\G$ of two individuals that are each a descendant of a different daughter population of population $p$.
%
% Because such a conflict implies that $P(\GT|\M) = 0$, we must also guarantee that $\Pref(\{\tau_p\}|\G)=0$.
%
Thus, the final step of constructing $\Pref(\GT|\Mref)$ is to construct $\Pref(\taus|\G,\thetas)=\Pref(\taus|\G)$ to have zero values whenever $P(\G|\M,\taus,\thetas)=0$.
%
This guarantee is achieved by computing for each $\tau_p$ an upper bound based on the coalescence events in $\G$
and defining $\Pref(\taus|\G)$ as a product of uniform distributions in the feasible ranges of $\taus$
%
(see  Appendix \ref{ap:cond_nomig} for complete derivation and proof).


\subsection*{Models with gene flow}

Assume now that the reference model is still the null model, $\M_0$, but the model of interest, $\M$, has a non-empty
set of migration bands, $B$, associated with migration rate parameters, $\migs=\{m_b:b\in B\}$.
%
Migration complicates the mapping between $\M$ and $\M_0$ because the genealogies in $\M$ hold information
about migration events, but the genealogies in $\M_0$ do not.
%
For a sequence of local genealogies $\G$ in $\M$, denote by $\Gc$ the coalescent trees implied by $\G$
and denote by $\Gm$ the information on migration events in $\G$ (locus, timing of event, branch in $\Gc$, source and target populations).
%
%Because the genealogies $\Gref$ in $\M_0$ have no migration events, then t
%There is a natural mapping between $\Gc$ (of $\M$) and $\Gref$ (of $\M_0$) and $P(\X|\Gref) = P(\X|\Gc)$.
Thus, a mapping between the hidden variables of $\M$ ($\Gc,\Gm,\T$) and the hidden variables of $\M_0$ ($\Gref,\theta_0$) can be defined by
mapping $\Gc$ to $\Gref$ and mapping some $\troot\in\T$ to $\theta_0$.
%
Consequently, the set of unmapped hidden variables is $\Z~=~ (\Gm,\taus,\migs,\thetas\setminus\{\troot\})$.
%
This implies a slight modification of the model-pairing conditional specified in Equation \ref{eq:pref_null}:
%
%
\begin{small}
\begin{align}
 \Pref(\GT|\M_0)
 &=~~ 
 %\Pref(\Gc,\troot,\Z|\M_0) ~~=~~
 P(\theta_0=\troot|\M_0)\  P(\Gref=\Gc|\M_0,\theta_0=\troot)\ \Pref(\Z|\Gc,\troot)  ~ .\label{eq:pref_mig}
\end{align}
\end{small}

The model-pairing condition of Equation \ref{eq:pref_integral} can be confirmed  by following a sequence of equalities similar to the ones we derived for 
the scenario without migration (see Equation \ref{eq:likelihood_null}).
%
We are thus left to specify the conditional distribution $\Pref(\Z|\Gc,\troot)$ to ensure that all $\GT$ for which $P(\Gc,\troot,\Z|\M,\X)=0$
also satisfy $\Pref(\Z|\Gc,\troot)=0$.
%
Since the genealogy trees $\Gc$ do not restrict the population size and migration rate parameters, we may define
the conditional probability for these parameters based on their prior probability under $\M$, so that their terms cancel out in the RBF ratio:
%
%
\begin{small}
\begin{align}
\frac{\Pref(\GT|\M_0) }{P(\GT|\M)}
&=~~ \frac{ P(\theta_0=\troot|\M_0) ~  P(\Gref=\Gc|\M_0,\theta_0=\troot) ~ \Pref(\Z|\Gc,\troot) } {P(\GT|\M)} \notag \\
&=~~ \frac{ P(\Gc|\M_0,\theta_0=\troot) }{ P(\Gc,\Gm|\M,\T)}~ 
     \frac{ P(\theta_0=\troot|\M_0) \prod_{p\neq root}\Pref(\theta_p)~\prod_{b}\Pref(m_b) }{P(\troot|\M)\prod_{p\neq\troot}P(\theta_p|\M)~\prod_{b}P(m_b|\M)}~
     \frac{ \Pref(\taus,\Gm|\Gc,\migs)}{P(\taus|\M)} \notag \\
&=~~ \frac{ P(\Gc|\M_0,\theta_0=\troot) }{ P(\Gc,\Gm|\M,\T)}~ 
     \frac{ P(\theta_0=\troot|\M_0)}{P(\troot|\M)}\
     \frac{ \Pref(\taus,\Gm|\Gc,\migs)}{P(\taus|\M)} ~. \label{eq:rbf_mig}
\end{align}
\end{small}

As in the case without migration, we are left to define the conditional probability distribution over the restricting hidden variables, which are in this case the divergence times
$\taus$ and the migration events $\Gm$.
%
The complex dependence between divergence times and migration events makes this particularly challenging.
%
For instance, a migration event between populations $p_1$ and $p_2$ at time $t$ implies that the divergence times of all populations ancestral to $p_1$ and $p_2$ is at least $t$,
%
but at the same time this migration event may also relax the upper bound of these divergence times.
%
Thus, bounds on divergence times cannot be determined solely based on $\Gc$, and the conditional $\Pref(\taus,\Gm|\Gc,\migs)$ cannot be factored into a product of
two separate probability distributions for $\taus$ and $\Gm$.
%
In Appendix \ref{ap:cond_mig} we present a specification for the joint conditional distribution $\Pref(\taus,\Gm|\Gc,\migs)$,
which addresses this complex dependence and ensures that $\Pref(\taus,\Gm|\Gc,\migs)=0$ whenever $P(\taus,\Gm,\Gc|\M)=0$.
%%
This construction results in additional terms canceling out with terms in  the genealogy likelihood
$P(\Gc,\Gm|\M,\T)$, to further reduce the variance of the RBF ratio.


\subsection*{The comb reference model}


The null model has the unique advantage of being a valid reference for the comparison of any two models.
This advantage, however, comes at the cost of collapsing all population structure.
%
In many cases we know the population designation of the sampled individuals, and model uncertainty is restricted to the relationships between the sampled populations.
%
To capture this simple structure we use a population phylogeny with a single ancestral population splitting simultaneously into all sampled populations.
We refer to such models as {\em comb} models and denote them by $\Mcomb$,  due to the comb-like structure of the population phylogeny.
%
A comb model is defined by: (1) a set of sampled (leaf) populations, $L$; (2) an ancestral population, $comb$; and (3) a set of migration bands $B_L$ between populations in $L$.
%
The resulting demographic model, $\Mcomb(L,B_L)$, has $|B_L|+|L|+2$ parameters: $\Tref ~=~ (\tacomb, \widetilde{\thetas},\widetilde{\migs})$,
where $\widetilde{\thetas}=\{\theta_p:p\in L\cup \{comb\}\}$ and $\widetilde{\migs} = \{m_b:b\in B_L\}$.
%


Consider a demographic model, $\M(\Tr,B)$, and its corresponding comb model, $\Mcomb(L,B_L)$, defined by $L=leaves(\Tr)$ and $B_L=B \cap (L \times L)$.
%
For brevity, we refer to $\Mcomb(L,B_L)$ simply as $\Mcomb$.
%
The model-pairing conditional distribution for $\M$ and $\Mcomb$ is constructed by first defining a mapping between the hidden variables of $\M$ ($\GT$) and the hidden variables of $\Mcomb$ ($\Gref\Tref$).
%
This mapping is derived from the requirement that below the comb divergence time ($\tacomb$) the comb model is identical to $\M$ and above it $\Mcomb$ is identical to the null model $\M_0$.
%
We thus set $\tacomb=\tmin\eqdef\min(\taus)$, to guarantee that all population divergence events in $\M$ map to the comb population in $\Mcomb$.
%
The migration rates of bands in $\B \cap (L \times L)$ and effective sizes of populations in $L$ are mapped into their counterparts in $\Tref$,
%
and following the mapping for the null model, a single ancestral population size parameter ($\troot$) is chosen to be mapped into $\thcomb$.
%
We denote the set of mapped migration rate and population size parameters of $\M$ collectively as $\Tcomb$.
%
Mapping between genealogies is obtained by {removing from $\G$ all migration events above time $\tmin$}.
The resulting collection of local genealogies are denoted by $\Gcomb$ and are directly mapped to $\Gref$.
%
The remaining unmapped hidden variables ($\Z$) of $\M$ consist of the following components:
\begin{enumerate}
 \item Unmapped population size parameters: $\{\theta_p : p\notin L\cup \{root\}\ \}$.
 \item Unmapped migration rate parameters:  $\{m_b: b\notin L \times L \}$.
 \item The identity of the ancestral population in $\Tr$ with minimum divergence time: $minAncPop=\argmin(\taus)$. 
   Note that this population may be \emph{any ancestral population with two leaf daughters}, and its identity is lost when mapping $\taus$ into $\tacomb$.
 \item The divergence times of all other populations: $\{\tau_p:p\neq minAncPop\}$.
 \item Information on all migration events in $\G$ above time $\tacomb$, which we denote by $\G_{m|>\tmin}$.
\end{enumerate}


A model-pairing conditional distribution for $\M$ and $\Mcomb$ is thus established by applying the mapping described above and
specifying a conditional distribution over the unmapped parameters, $\Pref(\Z|\Gcomb,\Tcomb,\tmin)$. The proof of the condition in Equation \ref{eq:pref_integral} is given below:
%
%
\begin{small}
\begin{align}
 \Pref(\GT|\Mcomb)
 &=
 P(\Tref=(\Tcomb,\tmin)|\Mcomb)\ P(\Gref=\Gcomb|\Mcomb,\Tcomb,\tmin)\ \Pref(\Z|\Gcomb,\Tcomb,\tmin)  ~ .\label{eq:pref_comb}\\
%\notag \\
%\end{align}
%\end{small}
%
%
%\begin{small}
%\begin{align}
P(\X|\Mcomb)
&= \int P(\Tref|\Mcomb)\ P(\Gref|\Mcomb,\Tref)\ P(\X|\Gref)\   d\Gref\Tref  \notag \\ %
&= \int P(\Tref=(\Tcomb,\tmin)|\Mcomb)\ P(\Gref=\Gcomb|\Mcomb,\Tcomb,\tmin)\ P(\X|\Gcomb)\  ~ d\Gcomb\Tcomb\tmin \notag \\ 
&= \int P(\Tref=(\Tcomb,\tmin)|\Mcomb)\ P(\Gref=\Gcomb|\Mcomb,\Tcomb,\tmin)\ P(\X|\Gcomb) \left( \int \Pref(\Z|\Gcomb,\Tcomb,\tmin) d\Z \right) d\Gcomb\Tcomb\tmin \notag \\ 
&= \int P(\Tref=(\Tcomb,\tmin)|\Mcomb)\ P(\Gref=\Gcomb|\Mcomb,\Tcomb,\tmin)\ \Pref(\Z|\Gcomb,\Tcomb,\tmin)\ P(\X|\G)\   d\GT \notag \\ 
&= \int \Pref(\GT|\M_0)\ P(\X|\G)\ d\GT ~. \label{eq:likelihood_comb} %\\
%\notag \\
\end{align}
\end{small}

The conditional distribution $\Pref(\Z|\Gcomb,\Tcomb,\tmin)$ is defined similar to its specification in the null model.
%
The unmapped population size and migration rate parameters are distributed according to their prior probability under $\M$ to eliminate terms in the RBF ratio.
%
The identity of the minimal ancestral population, $minAncPop$, is distributed uniformly among all ancestral populations in $\Tr$ with two leaf daughters.
%
We denote the number of such populations in $\Tr$ by $\kappa(\Tr)$.
%
The only unmapped variables restricted by $\Gcomb$ and $\tmin$ are the unmapped divergence times and migration events above time $\tmin$. Their conditional distribution,
$\Pref(\taus\setminus\{\tmin\},\G_{m|>\tmin}|\Gc,\migs)$, is defined using the process described for the null model (see Appendices \ref{ap:cond_nomig} and \ref{ap:cond_mig}).
%
This specification thus guarantees the condition of Equation \ref{eq:pref_support}, as in the case of the null reference model.
%
The resulting RBF ratio is expressed as follows:
%
%
\begin{small}
\begin{align}
\frac{\Pref(\GT|\Mcomb) }{P(\GT|\M)}
&= \frac{ P(\Tref=(\Tcomb,\tmin)|\Mcomb) ~ P(\Gref=\Gcomb|\Mcomb,\Tcomb,\tmin) ~ \Pref(\Z|\Gcomb,\Tcomb,\tmin) } {P(\GT|\M)} \notag \\
&= \frac{ P(\Gref=\Gcomb|\Mcomb,\Tcomb,\tmin) }{ P(\G|\M,\T)}~ 
   \frac{ P(\Tref=(\Tcomb,\tmin)|\Mcomb) }{ P(\Tcomb|\M) }~
   \frac{ \frac{1}{\kappa(\Tr)}\Pref(\taus\setminus\{\tmin\},\G_{m|>\tmin}|\Gc,\migs)}{P(\taus|\M)} ~. \label{eq:rbf_comb}
\end{align}
\end{small}


As in the case of the null reference model, the above RBF ratio has several terms canceling out. First, the conditional probabilities of the unmapped population size and
migration rate parameters cancel out with their priors under $\M$. Second, if we assume identical priors in both models for the mapped parameters,
then these cancel out as well in the second term of Equation \ref{eq:rbf_comb}.
Terms in the genealogy likelihood contributed by migration events above time $\tmin$ also cancel out in the ratio (see Appendix \ref{ap:cond_mig}).
Finally, the contribution of all events below time $\tmin$ (coalescence and migration) also cancel out.
If we denote the portion of $\G$ below time $\tmin$ by $\G_{<\tmin}$, and the portion above it by $\G_{>\tmin}$, then the contribution of $\G_{<\tmin}$ 
to the first term of the RBF ratio cancels out as follows:
%
%
\begin{small}
\begin{align}
\frac{ P(\Gcomb|\Mcomb,\Tcomb,\tmin) }{ P(\G|\M,\T)}
&=~~ \frac{ {P(\Gcomb}_{<\tmin}|\Mcomb,\Tcomb,\tmin) P({\Gcomb}_{>\tmin}|\Mcomb,\Tcomb,\tmin) }{ P(\G_{<\tmin}|\M,\T) P(\G_{>\tmin}|\M,\T)}   \notag \\
&=~~ \frac{ {P(\G}_{<\tmin}|\Mcomb,\Tcomb,~\tacomb=\tmin)}{ P(\G_{<\tmin}|\M,\Tcomb,~\min(\taus)=\tmin)} ~\frac{ P({\G}_{c|>\tmin}|\Mcomb,\thcomb=\troot) }{ P(\G_{>\tmin}|\M,\T)}   \notag \\
&=~~ \frac{ P({\G}_{c|>\tmin}|\M_0,\theta_0=\troot) }{ P(\G_{>\tmin}|\M,\T)} ~.  \label{eq:gen_ratio_comb}
\end{align}
\end{small}

The RBF may thus be re-expressed as follows:
%
%
\begin{small}
\begin{align}
\frac{\Pref(\GT|\Mcomb) }{P(\GT|\M)}
&= \frac{1}{\kappa(\Tr)} ~
   \frac{\Pref(\G_{>\tmin},\ \T\setminus\{\tmin\}\ |\ \M_0) }{P(\G_{>\tmin},\ \T\setminus\{\tmin\}\ |\ \M)} ~
   \frac{ P(\Tref\setminus\{\thcomb\}=(\Tcomb\setminus\{\troot\},\tmin)|\Mcomb) }{ P(\Tcomb\setminus\{\troot\}|\M) }~. \label{eq:rbf_comb1}
\end{align}
\end{small}


\subsection*{Collapsing clades}

In many cases of interest, the modeling uncertainty is restricted to a certain clade in the population phylogeny.
%
In such cases, we might wish to consider a reference model where a clade spanning a subset of the sampled populations $L$
is collapsed into a single population, or a comb model. The null and comb models discussed above are special cases where
$L$ is the entire set of sampled populations.
%
~~.\ .\ .\ ---TBA---

\subsection*{A general framework for defining the model pairing conditional distribution}

Previous sections presented specific types of models that may act as reference for model comparison.
%
In this section we described a general framework for defining a reference model and an appropriate model-pairing conditional
probability distribution.
%
Consider a model of interest $\M$ with hidden variables $\GT$ representing model parameters and genealogical information.
%
Denote by $\Om$  the domain of possible values for $\GT$.
%
Let $\Mref$ be a potential reference model with hidden variables $\Gref\Tref$ over domain $\Omref$ whose dimension is bounded from above by the
dimension of $\Om$.
%
A model-pairing conditional distribution can be defined for $\M$ and $\Mref$ by doing the following:
%
%
\begin{enumerate}
 \item Extend $\Gref\Tref$ by an additional set of hidden variables $\Zref$, s.t. the dimension of the extended domain $\Omref'$ for $\Gref\Tref\Zref$
 is the same as the dimension of $\Om$.
 \item \label{it:map} Define a bijective mapping $F$ from domain $\Om$ to domain $\Om'$.
 For a subset of hidden variables $\widetilde{U}\subset\Gref\Tref\Zref$, we denote by $F_{\widetilde{U}}$ the implied mapping into the sub-domain of $\widetilde{U}$.
 The mapping $F$ must preserve genealogical information in the sense that for every $\GT$, $P(\X|\G)=P(\X|F_{\Gref}(\GT))$.
  \item \label{it:cond} Define a conditional probability distribution $\Pref(\Zref|\Gref\Tref)$ that satisfies the following condition:
  %
  %
  \begin{small}
  \begin{equation}
  P(\GT|\M) = 0 ~\land~ P(F_{\Gref\Tref}(\GT)|\Mref) > 0 ~~~\Rightarrow~~~ \Pref(F_{\Zref}(\GT)|F_{\Gref\Tref}(\GT)) = 0 ~. \label{eq:pref-cond}
  \end{equation}
  \end{small}
\end{enumerate}

The model-pairing conditional $\Pref(\GT|\Mref)$ is defined using the conditional $\Pref(\Zref|\Gref\Tref)$
specified in (\ref{it:cond}) above, and the Jacobian determinant $J_F$ of the bijective mapping $F$ defined in (\ref{it:map})%
\footnote{The Jacobian $J_F$ is the determinant of the square matrix of partial derivatives, $\frac{\partial F_{\{\widetilde{u}\}}(\GT)}{\partial v}$,
where $v\in\GT$ and $\widetilde{u}\in\Gref\Tref\Zref$. Since $F$ is a bijection, then $J_F\neq 0$.}:
%
%
%
\begin{equation}\label{eq:pref_gen}
\Pref(\GT|\Mref) ~~=~~ P(F_{\Gref\Tref}(\GT)|\Mref)\ \Pref(F_{\Zref}(\GT)|F_{\Gref\Tref}(\GT))\ \frac{1} {J_F(\GT)}  ~.
\end{equation}
%
%

Because the conditional distribution $\Pref(\Zref|\Gref\Tref)$ satisfies Equation \ref{eq:pref-cond},
we are guaranteed that for all $\GT$ s.t. $P(\GT|\M)=0$ we also have $\Pref(\GT|\Mref)=0$.
%
Thus, to establish $\Pref$ as a model-pairing conditional probability distribution, we are left to show that $P(\X|\Mref) = \int P(\X|\G) \Pref(\GT|\Mref)\ d\GT$:
%
%
\begin{small}
\begin{align}
P(\X|\Mref)
&=~~ \int P(\X|\Gref)\ P(\Gref\Tref|\Mref)\ d\Gref\Tref  \notag \\ %
&=~~ \int P(\X|\Gref)\ P(\Gref\Tref|\Mref)  %
\left( \int \Pref(\Zref|\Gref\Tref)\ d\Zref\right)d\Gref\Tref \notag \\ 
&=~~ \int P(\X|\Gref)\ P(\Gref\Tref|\Mref)\ \Pref(\Zref|\Gref\Tref)\ d\Gref\Tref\Zref \notag \\ 
&=~~ \int P(\X|\G)\ P(F_{\Gref\Tref}(\GT)|\Mref)\ \Pref(F_{\Zref}(\GT)|F_{\Gref\Tref}(\GT))\ \frac{1} {J_F(\GT)}\  d\GT \notag \\ 
&=~~ \int P(\X|\G)\ \Pref(\GT|\Mref)\ d\GT ~. \label{eq:likelihood_gen} %
\end{align}
\end{small}

SOME DISCUSSION HERE~~.\ .\ .\ ---TBA---
\ignore{
%%%% ORIGINAL FORMULATION. SEE WHAT CAN STAY HERE  %%%%%
Assume first that we are interested in comparing our model of interest $\M$ to a reference model $\Mref$, which is a
{\em generalized} version of $\M$. This {\em generalization} is captured by a mapping $\Psi$ from the space of hidden
variables of $\M$ to the space of hidden variables of $\Mref$:~~ $\GTref = \Psi(\GT)$, such that if $P(\GT|\M)>0$,
then $P(\Psi(\GT)|\Mref)>0$. This implies that any instance of $\GT$ we sample in an MCMC given $\M$ will have positive
probability under the reference model (after mapping). Examples of generalizations and mappings are given later on,
but it is useful to think of a generalization as removing detail from the model (e.g., less population structure).
%
Our proposed importance sampling is based on the inverse map $\Psi^{-1}$, but the generalization mapping $\Psi$ will
typically not be invertible. This is because the reference model has less parameters and less detailed genealogical
information. To facilitate inversion, we append the reference model $\Mref$ with a series of additional hidden variables
($\Zref$), which are conditionally independent of the data $\X$ given $\GTref$. Consequently, $\Psi$ can be extended
an invertible mapping $\Psi'(\GT) = (\GTref,\Zref)$. We denote the Jacobian of this invertible mapping by $J_\Psi$,
and note that it will typically be equal to 1 (for most mappings that we will be discussing here).
%
Given the extended mapping $\Psi'$ and access to the posterior probability distribution of $\GT$ given $\X$ and
our assumed model $\M$, the {\em relative Bayes factor} can be estimated as follows:
}
\section*{Results}

-- TBA --

\section*{Discussion}
-- TBA --


% The bibtex filename
\renewcommand*{\refname}{Literature Cited}
\bibliographystyle{../../latex/bst/mbe}
\bibliography{../../latex/bib/compbio}
%

%=============================================================================

\clearpage{}
\section*{Tables}

%=============================================================================
\ignore{
\begin{table}[!h]
\caption{{\bf Model parameters}}
\noindent \begin{centering}
\begin{tabular}{lll}
\hline
{\bf Parameter} & {\bf Type} & {\bf Description}\\
\hline
\\[-2ex]
$\vect{\lambda^O} = \{\lambda^O_b\}_{b\in B}$ & neutral & Block-specific neutral scaling factor for the outgroup portion of\\
&& the phylogeny, used when computing the prior distributions for\\
&& the deep ancestral allele, $P(Z_i\ |\ O_i,\lambda^O_b)$\\
$\vect{\lambda} = \{\lambda_b\}_{b\in B}$ & neutral & Block-specific neutral scaling factor for divergence\\
$\vect{\theta} = \{\theta_b\}_{b\in B}$ & neutral & Block-specific neutral polymorphism rate\\
$\vect{\beta}=(\beta_1,\beta_2,\beta_3)$ & neutral & Relative frequencies
of the three derived allele frequency classes, $(0,f)$,\\
&& $[f,1-f]$, and $(1-f,1)$, within neutral polymorphic sites\\
$\rho$ & selection & Fraction of sites under selection within functional elements\\
$\eta$ & selection & Ratio of divergence rate at selected sites to local neutral\\
&& divergence rate\\
$\gamma$ & selection & Ratio of polymorphism rate at selected sites to local neutral\\
&& polymorphism rate\\
\hline
\end{tabular}
\par\end{centering}
\begin{flushleft}
\end{flushleft}
\label{tab:parameters}
\end{table}
}
%=============================================================================



%=============================================================================

\clearpage{}
\section*{Figure Legends}

\setlength{\parindent}{0pt} 
\setlength{\parskip}{2ex}

%=============================================================================

%=============================================================================
\ignore{

{\bf Figure \ref{fig:model-schematic}. Schematic description of \ins.}
%
The method measures the influence of natural selection by contrasting
patterns of polymorphism and divergence in a collection of genomic elements
of interest (gold) with those in flanking neutral sites (dark gray).
Nucleotide sites in both elements ($E_b$) and flanks ($F_b$) are grouped
into genomic blocks of a few kilobases in length ($b$) to accommodate
variation along the 
genome in mutation rates and coalescence time.  The model consists
of phylogenetic (gray), recent divergence (blue), and intraspecies
polymorphism (red) components, which are applied to genome sequences for
the target population ($X$, red) and outgroup species ($O$, gray).  At each
nucleotide position, the alleles at the most recent common ancestors of the
samples from the target population ($A$) and of the target population and
closest outgroup ($Z$) are represented as hidden variables and treated
probabilistically during inference.  The allele $Z$ determines whether or
not monomorphic sites are considered to be divergent (D).  Polymorphic
sites are classified as having low- (L) or high- (H) frequency derived
alleles based on $A$ and a frequency threshold $f$.  The labels shown here
are based on a likely setting of $Z$ and $A$.  Vertical ticks represent
single nucleotide variants relative to an arbitrary reference.  Inference
is based on differences in the patterns of polymorphism and divergence
expected at neutral and selected sites.
}
%=============================================================================

\clearpage{}
\section*{Figures}

%=============================================================================

%=============================================================================
\ignore{
\begin{figure}[h]
\begin{center}
%
\includegraphics[height=2.2in]{figures/insight_schematic.pdf}
%\includegraphics[height=1.7in]{figures/polydiv_data.pdf}
%
\caption{}
\label{fig:model-schematic}
%
\end{center}
\end{figure}
%=============================================================================
}
\clearpage


\appendix
\newcommand{\anc}{\geq_\Tr}
\newcommand{\nanc}{\ngeq_\Tr}
\section{\texorpdfstring{The conditional distribution $\Pref(\taus|\G)$ for models without migration}{Conditional distribution without migration}}\label{ap:cond_nomig}

When the hypothesis model $\M$ has no migration, its model-pairing conditional distribution with the null model $\M_0$ is determined by specifying a conditional
distribution for the divergence times, $\Pref(\taus|\G)$, such that $\Pref(\taus|\G)>0$ if and only if $P(\G|\taus,\M)>0$
(see Equations \ref{eq:pref_support} and \ref{eq:rbf_null}).
%
Let $(\Tr,\taus)$ be a timed population phylogeny and let $\G$ be a collection of coalescent trees in which every leaf is mapped to a leaf population in $\Tr$
and  every internal vertex $v$ corresponds to a coalescent event at time $t(v)$.
Then $P(\G|\taus,\M)>0$ if and only if the trees in $\G$ can be \emph{embedded} in  $(\Tr,\taus)$, as defined below.
%
\begin{definition}\label{def:embed}
 An embedding of a collection of local genealogies $\G$ in a timed population phylogeny $(\Tr,\taus)$ is a mapping, $pop:\G\rightarrow\Tr$,
 which satisfies the following conditions for every coalescence event $v\in\G$:
 \begin{enumerate}
  \item $pop(v)$ is alive at time $t(v)$:~~ $\tau(pop(v)) ~<~ t(v) ~\leq~ \tau(parent(pop(v)))$~.\label{cond:time}\\
  (if $p$ is a leaf population then $\tau(p)=0$ and if $p$ is the root population then $\tau(parent(p))=\infty$.)
  \item $pop(parent(v))$ is ancestral (or equal) to $pop(v)$:~~ $pop(parent(v)) \anc pop(v)$.\label{cond:anc}
 \end{enumerate}
\end{definition}

% Note that condition \ref{cond:anc} can be extended to show that for every $u$ ancestral to $v$, we have $pop(u) \anc pop(v)$.
%
%An interesting observation about embeddings is that they are unique.
%
Note that if $\G$ is embeddable in $(\Tr,\taus)$, then this embedding is unique, because given a coalescent event $v$ with daughter $u$,
there is only one population that is  alive at time $t(v)$ (condition \ref{cond:time}) and ancestral or equal to $pop(u)$ (condition \ref{cond:anc}).
%
A similar argument is used to establish a sufficient and necessary condition for embeddability below.
%
\begin{definition}[$mrcaPop$]\label{def:tmrca_pop}
 Given a coalescence event $v$ in a local genealogy whose leaves are assigned to the leaves of a population phylogeny $\Tr$,
 let ${mrcaPop(v)}$ denote the most recent common ancestor (MRCA) in $\Tr$ of all populations to which leaves in the subtree rooted at $v$ are mapped. %$\{pop(l):l\in leaves(v)\}$.
\end{definition}

\begin{lemma}\label{lem:embed}
 A collection of local genealogies $\G$ has an embedding in a timed population phylogeny $(\Tr,\taus)$ iff for every $v\in\G$ we have $t(v) > \tau(mrcaPop(v))$.
\end{lemma}
\begin{proof}
 ~\\
 $\Rightarrow$:~~ Consider an embedding $pop:\G\rightarrow\Tr$, and let $v$ be an arbitrary coalescence event in $\G$.
 Condition \ref{cond:anc} implies that $pop(v)\anc pop(l)$ for all leaves in the subtree rooted at $v$. % $l\in leaves(v)$
 We thus get $pop(v)\anc mrcaPop(v)$, and by condition \ref{cond:time}: $t(v) > \tau(pop(v)) \geq \tau(mrcaPop(v))$.\\
 %
 $\Leftarrow$:~~ Let $v$ be an arbitrary coalescence event in $\G$, and assume that $t(v) > \tau(mrcaPop(v))$. This means that there is a (unique) population, $p^*$,
 ancestral to $mrcaPop(v)$ that is also alive at time $t(v)$ (i.e., $\tau(p^*) ~<~ t(v)$\\
 $\leq~ \tau(parent(p^*))$). Define the embedding by mapping $v$ to population $p^*$.
 Condition \ref{cond:time} is guaranteed by construction. 
 Condition \ref{cond:anc} is proved by considering an arbitrary coalescence event $v$ and its parent $u=parent(v)$.
 Both $pop(u)$ and $pop(v)$ are ancestral (or equal) to $mrcaPop(v)$, because $mrcaPop(u)\anc mrcaPop(v)$.
 Thus either $pop(v)\anc pop(u)$ or $pop(u)\anc pop(v)$.
 Condition 1  implies that $pop(v)$ cannot be strictly ancestral to $pop(u)$ via the following sequence of inequalities:
 %
 $$\tau(parent(pop(u)) ~\geq~ t(u) ~>~ t(v) ~>~ \tau(pop(v))~.$$
 %
 Hence, $pop(u)\anc pop(v)$, establishing condition \ref{cond:anc}.
\end{proof}

Lemma \ref{lem:embed} directly implies a feasible range of every divergence time $\tau_p$:
%by noting that a given divergence time
%is feasible (i.e., $P(\G|\tau_p=\tau,\M)>0$) if and only if the values of the other divergence times can be set to allow the embedding of $\G$ in $(\Tr,\taus)$.

\begin{claim}\label{claim:tau_bound_nomig}
 Let $\G$ be a  collection of local genealogies whose leaves are mapped to leaves of a population phylogeny $\Tr$.
 Then for every ancestral population $p$, $P(\G|\tau_p=\tau,\M)>0$ iff $\tau \in [0 , ubound(p|\G) )$, where the upper bound of the feasible range for $\tau_p$
 is given by:
 %
 %
 \begin{small}
 \begin{align}
  ubound(p|\G)   =~~ & \min \{t(v):mrcaPop(v)\anc p\} \label{eq:ubound_nomig}  %\\
 \end{align}
 \end{small}
 %
 % 
\end{claim}

We thus define $\Pref(\taus|\G)$ as a product of uniform distributions for $\taus$ in their feasible ranges, as defined by Claim \ref{claim:tau_bound_nomig}.

\subsection*{Computing $\Pref(\taus|\G)$}

\textbf{TBA: explicit description of procedure for computing $ubound(p|G)$.}

\section{\texorpdfstring{The conditional distribution  $\Pref(\taus,\Gm|\Gc,\migs)$ for models with migration}{Conditional distribution with migration}}\label{ap:cond_mig}

As with the case without migration, the conditional distribution $\Pref(\taus,\Gm|\Gc,\migs)$ is constructed by first specifying the necessary and sufficient conditions
under which a genealogy with migration events $\G=(\Gc,\Gm)$ is embeddable in a timed population phylogeny $(\Tr,\taus)$.
%
Migration complicates these conditions because of two main reasons:
(1) migration breaks the fundamental assumption that genealogy branches move from a population to its parent in the phylogeny,
and (2) unlike coalescent events, migration events are mapped to specific populations and thus pose strict constraints on the embedding.
%
The first issue is addressed by examining \emph{migration-free} trees, obtained by cutting branches of the local genealogies in $\G$ at migration events.
%
We associate each migration event  $w\in\Gm$ with the branch in $\Gc$ on which it is placed, a specific time along that branch, a source population for migration, and a target population for migration.
%
Thus, each migration event, $w\in\Gm$, is a root of one migration-free tree mapped to population $target(w)$ and a leaf of another tree mapped to population $source(w)$.
~~[MAYBE ADD ILLUSTRATION OF THIS?]~~
%
In each migration-free tree, leaves are mapped to populations in $\Tr$ and branches move from a population to its parent, as assumed in condition \ref{cond:anc} of Definition \ref{def:embed}.
Hence, we can extend the operator $mrcaPop(v)$ of Definition \ref{def:tmrca_pop} as the MRCA of all populations to which the leaves of the migration-free subtree rooted at $v$ are mapped.
%
The following lemma specifies embeddability conditions based on this extended $mrcaPop$ operator and on the restriction that at the time of each migration event,
the source and target populations must be alive.


\begin{lemma}\label{lem:embed_mig}
 A collection of local genealogies $\G$  consisting of coalescent trees $\Gc$ and migration events $\Gm$ has an embedding in a timed population phylogeny $(\Tr,\taus)$ iff
 the following four conditions are satisfied:
 %
 %
 \begin{enumerate}
  \item \label{cond:ub_coal}  $\forall v\in \Gc: t(v) > \tau(mrcaPop(v))$
  \item \label{cond:mrca_mig} $\forall w\in\Gm: target(w)\anc mrcaPop(w)$
  \item \label{cond:ub_mig}   $\forall w\in\Gm: t(w) > \max(~\tau(source(w))~,~\tau(target(w))~)$
  \item \label{cond:lb_mig}   $\forall w\in\Gm: t(w) \leq \min(~\tau(parent(source(w)))~,~\tau(parent(target(w)))~)$
 \end{enumerate}
\end{lemma}

\begin{proof}
~\\
 %
 %
 $\Rightarrow$:~~ Assume a collection of local genealogies $\G$ embedded in a timed population phylogeny $(\Tr,\taus)$. For every coalescent event $v\in\Gc$, we know that
 $t(v)\geq \tau(pop(v))$, and $pop(v)\anc mrcaPop(v)$ (considering the migration-free tree that $v$ belongs to), implying condition \ref{cond:ub_coal}.
 Now consider an arbitrary migration event $w\in\Gm$, which is a root of some migration-free tree in $\G$ . Because this root is mapped to population $target(w)$, we get that
 $target(w)\anc  mrcaPop(w)$ (condition \ref{cond:mrca_mig}).
 Finally, conditions \ref{cond:ub_mig} and \ref{cond:lb_mig} are implied by the fact that $w$ is mapped to populations $target(w)$ (as the root of a migration-free tree) and
 $source(w)$ (as a leaf of a migration-free tree).\\
 %
 %
 $\Leftarrow$:~~ Assume a collection of local genealogies $\G$ and a timed population phylogeny $(\Tr,\taus)$ satisfying the four conditions of the lemma.
 We embed $\G$ in $(\Tr,\taus)$ by mapping every coalescent event to the population ancestral to $mrcaPop(v)$ that is also alive at time $t(v)$.
 This is the same mapping used in the proof of Lemma \ref{lem:embed}, when no migration was assumed, and as in that case,
 we can show that such a population exists (through condition \ref{cond:ub_coal}) and that for each coalescent event $v$  we have $pop(parent(v))\anc pop(v)$.
 Hence, the two conditions of Definition \ref{def:embed} are satisfied for all coalescent events.
 The same holds for all migration events, because
 conditions \ref{cond:ub_mig} and \ref{cond:lb_mig} imply that each migration event $w$ is mapped to source and target populations that are both alive at time $t(w)$,
 and condition \ref{cond:mrca_mig} implies that $target(w)$ is ancestral to the population to which the event at the bottom of the branch below $w$ is mapped.
 Thus the mapping satisfies the two conditions of Definition \ref{def:embed} with respect to all migration-free trees in $\G$, implying that $\G$ is embeddable
 in $(\Tr,\taus)$.
\end{proof}

Note that condition \ref{cond:mrca_mig} of the lemma specifies constraints on migration events in $\Gm$ and conditions \ref{cond:ub_coal}, \ref{cond:ub_mig}, and \ref{cond:lb_mig}
define the feasible range for divergence times, as defined below.
%
%
\begin{claim}\label{claim:tau_bound_mig}
 Let $\G$ be a  collection of local genealogies with migration events. Then for every ancestral population $p$, $P(\G|\tau_p=\tau,\M)>0$ iff for every $w\in\Gm$ we have $target(w)\anc mrcaPop(w)$
 and   $\tau \in [lbound(p|\G) , ubound(p|\G) )$, where the bounds of the feasible range for $\tau_p$
 are given by:
 %
 %
 \begin{small}
 \begin{align}
  lbound(p|\G)   =~~ & \max\left\{t(w)| w\in \Gm \wedge ( p\anc parent(source(w)) \vee p\anc parent(target(w))) \right\} \label{eq:lbound_mig}\\
  ubound(p|\G)   =~~ & \min(ubound_1(p|\G) , ubound_2(p|\G)) \label{eq:ubound_mig}  \\
  ubound_1(p|\G) =~~ & \min\left\{t(v)| v\in \Gc \wedge mrcaPop(v)\anc p\right\} \label{eq:ubound1_mig}\\
  ubound_2(p|\G) =~~ & \min\left\{t(w)| w\in \Gm \wedge ( source(w)\anc p \vee target(w)\anc p) \right\} \label{eq:ubound2_mig} %\\
 \end{align}
 \end{small}
 %
 % 
\end{claim}

We thus define the conditional distribution $\Pref(\Gm,\taus|\Gc,\migs)=\Pref(\taus|\G)\Pref(\Gm|\Gc,\migs)$,
where $\Pref(\taus|\G)$ is the product of uniform distributions for $\taus$ in their feasible ranges, as defined by Claim \ref{claim:tau_bound_mig},
and $\Pref(\Gm|\Gc,\migs)$ is defined using a probabilistic protocol for sampling migration events.
%
This protocol mimics the true migration model of $\M$ as much as possible without knowing the divergence times.
% while ensuring that the resulting collection of genealogies, $\G$, is embeddable in some timed version of the population phylogeny (i.e., there exist $\taus$ s.t. $P(\G|\taus,\migs,\M)>0$).
%
Migration events are sampled backward in time by holding for each branch $(u,v)\in\Gc$ the set of populations it may be embedded in (those ancestral to $mrcaPop(v)$),
and allowing the branch to migrate back along any migration band whose target population is one of those populations.
%
The protocol starts by enabling migration in all bands, and it removes a migration band $b$ from consideration when the protocol reaches time
$t=\min\left(ubound(parent(source(b))|\G),ubound(parent(target(b))|\G)\right)$, as defined by Equations \ref{eq:ubound_mig}-\ref{eq:ubound2_mig}.
%
By doing this, the protocol ensures that the resulting $\G$ will be embeddable in some timed version of the population phylogeny (see Claim \ref{claim:protocol} below).

\noindent \textbf{Sampling protocol for } $\Pref(\Gm|\Gc,\migs)$:
%
%
\begin{enumerate}\vspace{-1em}
  \item \textbf{Initialization:}
  \begin{enumerate}
    \item \label{step:init_mapping} Initialize set of living branches: $E_{live}\leftarrow\{(u,v)\in E(\Gc)| v \text{ is a leaf}\}$. Map each $(u,v)\in E_{live}$ to the sampling population of the leaf $v$
    and all populations ancestral to it: $pops((u,v))\leftarrow\{p|p\anc pop(v)\}$.
    \item Initialize living migration bands: $B_{live}\leftarrow B$.
    \item Initialize time: $t\leftarrow 0$.
 \end{enumerate}
 
  \item \label{step:interval} \textbf{Determine current migration rates:}
    Determine the number of branches currently mapped to each population, $n[p]=|\{e\in E_{live}:p\in pops(e)\}|$, and 
    compute the \emph{effective rate} of each living migration band:  $\lambda[b] = m_b\times n[target(b)]$
    (the migration rate scaled by the number of potentially migrating branches).
    %If $0 = \lambda=\sum_{b\in B_{live}}\lambda[b]$, then terminate scan (no more migration events).
  \item  \label{step:deltaT} \textbf{Sample time of next migration:} 
    Sample a waiting time $\Delta t$ for the next migration event according to an exponential distribution with rate $\lambda=\sum_{b\in B_{live}}\lambda[b]$. 
    If there are no live migration bands with positive rates, then $\lambda=0$ and the scan terminates (no more migration events to sample).
    Otherwise, set $t\leftarrow t+\Delta t$ and compare $t$ to the time of the next coalescent event back in time, $v$.

  \item \label{step:sample-mig} If $t<t(v)$, then \textbf{sample migration event:}
  \begin{enumerate}
    \item \label{step:mig_band} Sample a migration band $b\in B_{live}$ using a categorical distribution with $p_b=\frac{\lambda[b]}{\lambda}$.
    \item \label{step:mig_branch} Select a branch for migration $e\in E_{live}$ uniformly at random among the $n[target(b)]$ branches mapped to the target population of the selected migration band.
    \item Add a new migration event $w$ to $\Gm$ on branch $e$ from population $source(b)$ to population $target(b)$ at time $t$.
    \item \label{step:mig_mapping} Update the population mapping of edge $e$: $pops(e)\leftarrow\{p:p\anc source(b)\}$.
    %\item Update all $n[p]$ and $\lambda[b]$ values affected by the change of population mapping for $l$.
    %, and update the effective migration rates of all migration bands in $B_{live}$ whose target population was affected by this change.
    % \item \label{step:lb} Ensure that the parents of $source(b)$ and $target(b)$ do not diverge below the sampled migration band by updating lower bounds:
    %  $lBound(\tau_{parent(source(b))}), lBound(\tau_{parent(target(b))}) \leftarrow t$.
    % \item \label{step:ub_mig} Ensure that $source(b)$ and $target(b)$ diverge below the sampled migration event, by
    %   setting $uBound(\tau_{source(b)}),uBound(\tau_{source(b)})\leftarrow t$ (if it has not already been set), and removing
    \item  \label{step:ub_mig} Remove from $B_{live}$ all migration bands whose source or target population is a strict descendant of either $source(b)$ or $target(p)$.
    Formally, remove band $b'$ iff there is $p'\in\{source(b'),target(b')\}$ and $p\in\{source(b),target(b)\}$  s.t. $p\anc parent(p')$.
   
    \item Go to Step \ref{step:interval}.
  \end{enumerate}
  
  \item If $t \geq t(v)$, then \textbf{encounter coalescence event:}
  \begin{enumerate}
    \item \label{step:coal} Let $e_1$ and $e_2$ be the two branches coalescing in $v$, and let $e$ be the branch above $v$.
    \item Update current branches: $\E_{live}\leftarrow E_{live}\setminus\{e_1,e_2\}\cup\{e\}$.
    \item \label{step:coal_mapping} Map the new branch: $pops(e) = pops(e_1)\cap pops(e_2)$.
    %\item Decrease $n[p]$ by one for all populations in $pops(l_1)\cup pops(l_2)$ and updated $\lambda[b]$ values of migration bands whose target population is in this union. 
    \item \label{step:ub_coal} Remove from $B_{live}$ all migration bands whose source or target is a strict descendant of the most recent population in $pops(e)$.
    Formally, if $p_0$ is the most recent population in $pops(e)$, then remove band $b$ iff $p_0 \anc parent(source(b))$ or $p_0\anc parent(target(b))$.
    \item Set $t\leftarrow t(v)$ and go to Step \ref{step:interval}.
  \end{enumerate}
  
  % \item \label{step:sample-taus}  \textbf{Sample divergence time parameters:}
  % \begin{enumerate}
  %   \item Propagate lower bounds for divergence times up $\Tr$, such that $lBound(\tau_p)\leq lBound(\tau_{parent(p)})$.
  %  \item Sample a value for $\tau_p$ uniformly at random in the interval $[lBound(\tau_{parent(p)}),uBound(\tau_{parent(p)})]$.
  % \end{enumerate}
\end{enumerate}

The following claim establishes the validity and completeness of the above protocol for $\Pref(\Gm|\Gc,\migs)$:
%
%
\begin{claim}\label{claim:protocol}
 %The protocol described above samples a set of migration events $\Gm$ given a collection of coalescent trees $\Gc$ and migration rates $\migs$
 $\Pref(\Gm|\Gc,\migs)>0$ ~~ iff ~~there exist $\taus$ s.t. $P(\Gc,\Gm|\taus,\migs,\M)>0$.
\end{claim}
%
%
\begin{proof}
 First, note that the protocol maps each branch $(u,v)$ to the set of populations ancestral to $mrcaPop(v)$: $pops((u,v))=\{p:p\anc mrcaPop(v)\}$.
 This is done by the appropriate initialization of the mapping in leaf branches in step \ref{step:init_mapping} and branches above migration events in step \ref{step:mig_mapping},
 and by the appropriate intersection update in branches above coalescent events in step \ref{step:coal_mapping}.
 Both directions of the claim are now proved using this observation and the conditions of Claim \ref{claim:tau_bound_mig}\\
 %
 %
 $\Rightarrow$\\
 % 
 Let $\Gm$ be the set of migration events sampled by the protocol given $\Gc$ and $\migs$.
 To establish that there exist $\taus$ s.t. $P(\Gc,\Gm|\taus,\migs,\M)>0$ using Claim \ref{claim:tau_bound_mig}, we need to show that:
 (1) every sampled migration event in $\Gm$ satisfies  $target(w)\anc mrcaPop(w)$, and
 (2) the resulting $\G$ satisfies $lbound(p|\G)<ubound(p|\G)$ for every ancestral population $p$.
 Let $w\in\Gm$ be an arbitrary migration event and denote by $e(w)$ the branch in $\Gc$ on which $w$ is sampled.
 Then, $target(w)\subseteq pops(e(w))$ (step \ref{step:mig_branch}), implying that $target(w)\anc mrcaPop(w)$, as required by Claim \ref{claim:tau_bound_mig}.
 Now, consider an arbitrary ancestral population $p$, and denote for brevity $lb=lbound(p|\G)$, $ub_1=ubound_1(p|\G)$, and $ub_2=ubound_2(p|\G)$ (Equations \ref{eq:lbound_mig}-\ref{eq:ubound2_mig}).
 We will show that $lb<\min(ub_1,ub_2) = ubound(p|\G)$.
 
 Let $v$ be the coalescent event realizing $ub_1$ and let $w$ and $w'$ be the migration events realizing $ub_2$ and $lb$, respectively.
 Note that if one of these events does not exist, then the appropriate bound is set to its extreme value (0 for $lb$ and $\infty$ for $ub_1$ and $ub_2$),
 and the inequality above holds. Otherwise,  the definition of $w'$ and $lb$ implies that either $p\anc parent(source(w'))$ or $p\anc parent(target(w'))$,
 and the definition of $w$ and $ub_2$ implies that either $source(w)\anc p$ or $target(w)\anc p$.
 %
 Hence, the condition of step \ref{step:ub_mig} of the protocol is satisfied for the migration band of event $w'$ ($b'$) when the protocol samples event $w$.
 This means that migration band $b'$ is not alive after sampling $w$ and $lb = t(w') < t(w) = ub_2$.
 Similarly, the condition of step \ref{step:ub_coal} of the protocol is satisfied for migration band $b'$ when the protocol encounters coalescent event $v$
 ($p_0=mrcaPop(v)$). Hence, migration band $b'$ is not alive after encountering $v$ and $lb = t(w') < t(v) = ub_1$, completing the requirements of Claim \ref{claim:tau_bound_mig}.\\
 %
 %
 $\Leftarrow$\\
 % 
 Let $(\G,\taus)$ be a collection of local genealogies and divergence times s.t. $P(\Gc,\Gm|\taus,\migs,\M)>0$.
 We will show that the migration events in $\Gm$ can be sampled by the protocol (with some positive probability).
 Consider an arbitrary migration event $w\in\Gm$ and assume that the protocol reached time $t(w)$ in $\Gc$ after
 having correctly sampled all events $w'\in\Gm$ s.t. $t(w')<t(w)$.
 To prove that event $w$ can be sampled with positive probability we need to establish that:
 (1) its migration band $(p_s,p_t)=(source(w),target(w))$ is alive at time $t(w)$,
 and (2) its branch, $e$, is mapped to the target population $p_t$.
 The second requirement follows from Claim \ref{claim:tau_bound_mig}, which implies that $p_t\anc mrcaPop(w)$, and our observation on the mapping that
 states that each branch is mapped to the set of populations ancestral to its $mrcaPop$.
 
 To establish the first requirement we need to prove that migration band $b=(p_s,p_t)$ was not removed from $B_{live}$ before time $t(w)$.
 The protocol removes migration bands from $B_{live}$ either after sampling migration events (step \ref{step:ub_mig}) or
 after encountering a coalescent events (step \ref{step:ub_coal}).
 Let $w'\in\Gm$ be an arbitrary migration event sampled before $w$ s.t. $t(w')<t(w)$.
 Claim \ref{claim:tau_bound_mig} implies that for $p'\in\{source(w'),target(w')\}$ we have $\tau(p')<ubound_2(p'|\G)\leq t(w') < t(w)$, and
 for $p\in\{p_s,p_t\}$ we have $t(w)\leq lbound(parent(p)|\G) \leq \tau(parent(p))$.
 Hence, $\tau(p')<\tau(parent(p))$, implying that populations $source(w')$ and $target(w')$ are not strictly ancestral to populations $p_s$ and $p_t$,
 and so the migration band $(p_s,p_t)$ is not removed from $B_{live}$ after sampling event $w'$ (see step \ref{step:ub_mig}).
 
 Now, let $v\in\Gc$ be an arbitrary coalescent event encountered before sampling $w$ s.t. $t(v)<t(w)$.
 Claim \ref{claim:tau_bound_mig} implies that for $p'=mrcaPop(v)$ we have $\tau(p')<ubound_1(p'|\G) \leq t(v)<t(w)$ and
 for $p\in\{p_s,p_t\}$ we have $t(w)\leq lbound(parent(p)|\G) \leq \tau(parent(p))$.
 This means that $\tau(p')<\tau(parent(p))$, implying that population $mrcaPop(v)$ is not strictly ancestral to populations $p_s$ and $p_t$,
 and so the migration band $(p_s,p_t)$ is not removed from $B_{live}$ after encountering event $v$ (see step \ref{step:ub_coal}).
 Thus, migration band $(p_s,p_t)$ is alive at time $t=t(w)$, and the branch $e$ is mapped to $p_t$,
 allowing the protocol to sample $w$ at time $t(w)$ with positive probability.
 \end{proof}

\subsection*{Computing the conditional probability}

NEED TO TIGHTEN DESCRIPTION HERE AND ADD DETAILS.\\

Now that we have fully defined the conditional probability distribution $\Pref(\Gm,\taus|\G,\migs)$, we turn to describe how to compute it for given values of $(\G,\taus,\migs)$.
The divergence time conditionals, $\Pref(\taus|\G)$, are defined as a product of uniform distributions in the feasible space of every parameter, as defined by Claim \ref{claim:tau_bound_mig}.
The bounds $lbound$ and $ubound_2$ are easy to compute by traversing all migration events in $\Gm$, and the bound $ubound_1$ can be computed by recursively computing $mrcaPop$ for all coalescent
events in $\Gc$, as described in the previous section. This is done by considering the migration-free trees defined by $\G$.
%
The conditional probability for the migration events, $\Pref(\Gm|\Gc,\migs)$ is computed according to the sampling protocol described above.
%
As in a standard model of migration at constant rate, this probability can be expressed as a product of contributions across migration bands [ADD REF TO RELEVANT EQUATION?]:
\begin{equation}\label{eq:pref-m}
 \ln \left( \Pref(\Gm|\Gc,\migs) \right) ~=~ \sum_b \left( \ln( m_b) \cdot numMigs(\Gm,b)^{m_b} -m_b \cdot \widetilde{migStats}(\G,b) \right) ~.
\end{equation}

Consequently, the contribution of migration band $b$ to $\Pref(\Gm|\Gc,\migs)$ is very similar to its contribution to $P(\G|\T,\M)$, and the ratio
between these contributions is defined by the difference between $migStats(\G,b)$ and $\widetilde{migStats}(\G,b)$.
Both migration statistics are defined as sum across time intervals in population $target(b)$ across the life span of the migration band.
In model $\M$, the life span starts at $t=\max(\tau(source(b)),\tau(target(b)))$ and ends at $t=\min(\tau(parent(source(b))),\tau(parent(target(b))))$.
In the sampling protocol the life span starts at time $t=0$ and ends at $t=\min(ubound(parent(source(b))|\G),ubound(parent(target(b))|\G))$.
Note that the life span in $\M$ is contained in the protocol life span, and in this time the lineages mapped to population $target(b)$ are the same in both cases.
Thus the residual difference, $migStats(\G,b)-\widetilde{migStats}(\G,b)$, is computed by considering intervals mapped to $target(b)$ in the protocol and not in $\M$.
For instance, if $b$ is a migration band between two sampled populations, then its life span in $\M$ and in the protocol starts at $t=0$, and the residual is computed by 
determining which branches of $\G$ are mapped to population $target(p)$ in the time interval between $t=\min(\tau(parent(source(b))),\tau(parent(target(b))))$ and
$t=\min(ubound(parent(source(b))|\G),ubound(parent(target(b))|\G))$.



\end{document}
